File: sparseMRA-main/Gen_Fig1.m
% This script generates Figure 1 of the paper "SPARSE MULTI-REFERENCE
% ALIGNMENT: SAMPLE COMPLEXITY AND COMPUTATIONAL HARDNESS" by Tamir Bendory, Oscar Mickelin, and Amit Singer
%
% Last update: September 20, 2021
% Tamir Bendory

clear;
close all;
clc;

%% parameters

rng(987); % seed

N = 80; % signal's length
% N = 120;
K_vec = 5:17;  % sparsity level
num_iter = 500;
% last_iter_rrr_120 = zeros(length(K_vec), num_iter);
last_iter_rrr_80 = zeros(length(K_vec), num_iter);

%% RRR parameters

parameters.beta = 1/2;
parameters.max_iter = 10^6;
parameters.verbosity = 1;
parameters.th = 1e-5;

%% Main loop

for k = 1:length(K_vec)
    K = K_vec(k);
    %%
    for iter = 1:num_iter
        
        %% generating a binary signal with random support
        ind_true = randperm(N);
        ind_true = ind_true(1:K);
        x_true = zeros(N,1);
        x_true (ind_true) = 1;
        
        %% measurements
        
        y = abs(fft(x_true)).^2;
        
        %% RRR
        x_init = randn(N, 1);
        tic
        [x_est, error, diff, last_iter_rrr_80(k,iter)] = RRR(sqrt(y), x_init, K, parameters);
    end
    fprintf('K = %g\n', K)
%     save('last_iter_rrr_120','last_iter_rrr_120');
end

%% plotting

fig = figure;
hold on;
% plot(K_vec, mean(last_iter_rrr, 2));
plot(K_vec, median(last_iter_rrr_80, 2), 'linewidth',1);
plot(K_vec, median(last_iter_rrr_120, 2), 'linewidth',1);
legend({'L=80', 'L=120'})
set(gca, 'YScale', 'log')
ylabel('iterations')
xlabel('K')
axis tight;
%pdf_print_code(gcf, 'RRR_iterations_80.pdf', 11)
saveas(fig,'Fig1.png')


%%

% 
%     load('last_iter_rrr_80.mat');
%     load('last_iter_rrr_120.mat');
%     K_vec = 5:17;  % sparsity level
%     
%     figure;
%     hold on;
%     plot(K_vec, median(last_iter_rrr_80, 2), 'linewidth',1);
%     plot(K_vec, median(last_iter_rrr_120, 2), 'linewidth',1);
%     set(gca, 'YScale', 'log')
%     ylabel('iterations (log scale)')
%     xlabel('M')
%     axis tight;
%     legend('L=80', 'L=120', 'Location', 'northwest');
%     pdf_print_code(gcf, 'RRR_XP.pdf', 11)
%     
    

File: Translate/Gen_Fig1.py
import numpy as np
from matplotlib import pyplot as plt
from RRR import RRR
import os, scipy

rand_seed = np.random.randint(10000)
print(f'random seed: {rand_seed}')
np.random.seed(rand_seed)


def generate_last_itervec(N=80):
    K_vec = np.arange(5, 18)
    num_iter = 500

    last_iter_rrr = np.zeros((len(K_vec), num_iter))

    parameters = {
        'beta': 1/2,
        'max_iter': int(1e6),
        'verbosity': 0,
        'th': 1e-5
    }

    for k in range(len(K_vec)):
        K = K_vec[k]
        print(f'K = {K:g} \n')

        for _iter in range(num_iter):
            ind_true = np.random.permutation(N)
            ind_true = ind_true[1:K]
            x_true = np.zeros((N, 1), dtype=int)
            x_true[ind_true] = 1

            y = np.abs(np.fft.fft(x_true, axis=0)) ** 2

            x_init = np.random.randn(N, 1)
            x_est, error, diff, last_iter_rrr[k, _iter] = RRR(np.sqrt(y), x_init, K, parameters)

        print(f'K = {K:g} \n')

    return K_vec, last_iter_rrr


if __name__ == '__main__':
    K_vec, last_iter_rrr_80 = generate_last_itervec(N=80)
    _, last_iter_rrr_120 = generate_last_itervec(N=120)

    med_last_iter_rrr_80 = np.median(last_iter_rrr_80, axis=1)
    med_last_iter_rrr_120 = np.median(last_iter_rrr_120, axis=1)

    fig = plt.figure(figsize=[6.4 * 1.5, 4.8 * 1.5])
    plt.plot(K_vec, med_last_iter_rrr_80)
    plt.plot(K_vec, med_last_iter_rrr_120)
    plt.yscale('log')
    plt.ylabel('iterations')
    plt.xlabel('K')
    fig.tight_layout()
    plt.legend(['L=80', 'L=120'])

    plt.savefig('Fig1.png')
